Objective-C/CBLDatabase.mm

```diff
@@ -1214,6 +1176,18 @@ - (void) safeBlock:(void (^)())block {
     }
 }
 
+- (CBLCollection*) mustDefaultCollection: (NSError**)outError {
+    NSError* e = nil;
+    CBLCollection* c = [self defaultCollection: &e];
+    if (!c) {
+        [NSException raise: NSInternalInconsistencyException
+                    format: @"Attempt to perform an operation on removed collection %@", e];
+        if (outError)
+            *outError = e;
+    }
+    return c;
+}
+
```

Comment:
How about this?

```
- (instancetype) initWithName: (NSString*)name
                       config: (nullable CBLDatabaseConfiguration*)config
                        error: (NSError**)outError {
    ...
    if (self) {
        ....
        NSError* error;
        _defaultCollection = [self collectionWithName: kCBLDefaultCollectionName scope: kCBLDefaultScopeName error: &error];
        if (error) { 
           CBLWarn(Database, @"%@ : Error getting the default collection: %@", self, error); 
       }
    }
    return self;
}

- (nullable CBLCollection*) defaultCollection: (NSError**)error {
   CBL_LOCK(self) {
       if (![self mustBeOpen: error]) {
           return nil;
       }
       return _defaultCollection.isValid ? _defaultCollection : nil;
   }
}

- (CBLCollection*) defaultCollectionOrThrow {
    if (!_defaultCollection )
        throwNotOpen();
    return _defaultCollection; 
}

- (nullable NSDate*) getDocumentExpirationWithID: (NSString*)documentID {
    NSError * error;
    NSDate* exp = [[self defaultCollectionOrThrow] getDocumentExpirationWithID: documentID error: error]];
    throwIfNotOpenError(error);
    return exp;
}

static void throwNotOpen() {
    [NSException raise: NSInternalInconsistencyException format: @"The database was closed, or the default collection was deleted."];
}

static void throwIfNotOpenError(NSError* error) {
    if (error && error.domain == CBLErrorDomain && error.code == CBLErrorNotOpen) {
        throwNotOpen();
    }
}
```

Reply:
Note : defaultCollectionOrThrow() doesn't check if the collection is valid because it will be checked when calling any [CBLCollection XXX] method, and a big plus, the method doesn't require locking.